### ISSUE TITLE
Trivy operator transfer data a lot through NAT gateway

### ISSUE THREAD

**Issue by chary1112004** [aquasecurity/trivy-operator#1744]
**What steps did you take and what happened:**

Hi,

Currently we detect trivy operator transfer data through NAT gateway (we are using AWS) quite amount data then it really cost.

Helm chart installed: https://artifacthub.io/packages/helm/trivy-operator/trivy-operator/0.3.0

Any suggestion to resolve this issue would be very much appreciated!

Thanks

**Environment:**

- Trivy-Operator version (use `trivy-operator version`): 0.3.0
- Kubernetes version (use `kubectl version`): 1.25
- OS (macOS 10.15, Windows 10, Ubuntu 19.10 etc):

**Comment by chen-keinan** [comment]
@chary1112004 Trivy-operator v0.18.0 use SBOM cache and reduce the amount of download of image scan.

**Comment by chary1112004** [comment]
@chen-keinan thank you for suggestion. We will upgrade to v0.18.0 and verify it. Thanks!

**Comment by chary1112004** [comment]
@chen-keinan I would like to have question, what happen for images that are already downloaded and running in kubernetes during scanning. Does it download again to scan? Thanks!

**Comment by chen-keinan** [comment]
> @chen-keinan I would like to have question, what happen for images that are already downloaded and running in kubernetes during scanning. Does trivy download image inside before it starts scanning? Thanks!

its depend on which node the images where downloaded , if its on the same node then trivy will re-use.
we added sbom cache to re-use prev. scans of trivy  it should solve the images download issue.

Note: if you use [filesystem scan command](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L466) then images will not be downloaded however it require privileged access

**Comment by chen-keinan** [comment]
@chary1112004 do you still want to keep this issue open ?

**Comment by chary1112004** [comment]
@chen-keinan I just come back this to verify. I will keep update once it works for us so please keep it open for now! Thanks!

**Comment by aarontavio** [comment]
Hi @chen-keinan, I work with @chary1112004.

I was looking for documentation about the `trivy.command` setting in the context of the trivy operator. This is what I found:

- trivy operator docs: https://aquasecurity.github.io/trivy-operator/v0.18.1/docs/vulnerability-scanning/trivy/
- trivy scanner docs: https://aquasecurity.github.io/trivy/v0.48/docs/target/filesystem/

If I run trivy locally then the filesystem scan command is clear to me, but if I set `trivy.command` to `filesystem` what will it happen? will it scan the pod's filesystem? and with `rootfs` will trivy scan the kubernetes node's filesystem?

Can you provide some clarification on this? The documentation is not clear enough I think.

Thanks in advance.

**Comment by chen-keinan** [comment]
@aarontavio using `trivy.command. = filesystem` is scanning the same image but without downloading it.
 it use the downloaded image on the node already. but it required higher privileges.
 
 see here the differences between [fs to rootfs support](https://aquasecurity.github.io/trivy/v0.24.4/vulnerability/detection/language/)

**Comment by chen-keinan** [comment]
> @chen-keinan could you please give us more explain about trivy.command = rootfs with filesystem and image since when we use trivy.command. = filesystem and it generates issue in #1792? Thanks!

command = Image:  download the image from registry
comand = filesystem is scan the local downloaded image on node

**Comment by chary1112004** [comment]
> > @chen-keinan could you please give us more explain about trivy.command = rootfs with filesystem and image since when we use trivy.command. = filesystem and it generates issue in #1792? Thanks!
> 
> command = Image: download the image from registry comand = filesystem is scan the local downloaded image on node

Yes, you explain about **image** and **filesystem** before. I would like to ask about option **rootfs**.
trivy.command = rootfs

**Comment by chen-keinan** [comment]
> @aarontavio using `trivy.command. = filesystem` is scanning the same image but without downloading it. it use the downloaded image on the node already. but it required higher privileges.
> 
> see here the differences between [fs vs. rootfs support](https://aquasecurity.github.io/trivy/v0.24.4/vulnerability/detection/language/)

@chary1112004  see this comment

**Comment by chen-keinan** [comment]
> > > @aarontavio using `trivy.command. = filesystem` is scanning the same image but without downloading it. it use the downloaded image on the node already. but it required higher privileges.
> > > see here the differences between [fs vs. rootfs support](https://aquasecurity.github.io/trivy/v0.24.4/vulnerability/detection/language/)
> > 
> > 
> > @chary1112004 see this comment
> 
> @chen-keinan I could see the rootfs also scan the same what image scan. What I would like to ask that is **rootfs** download the image from registry (same as **image**) or scan the local downloaded image on node (same as **filesystem**). Thanks!

`rootfs` is working the same as `filesystem`

**Comment by chary1112004** [comment]
> > > > @aarontavio using `trivy.command. = filesystem` is scanning the same image but without downloading it. it use the downloaded image on the node already. but it required higher privileges.
> > > > see here the differences between [fs vs. rootfs support](https://aquasecurity.github.io/trivy/v0.24.4/vulnerability/detection/language/)
> > > 
> > > 
> > > @chary1112004 see this comment
> > 
> > 
> > @chen-keinan I could see the rootfs also scan the same what image scan. What I would like to ask that is **rootfs** download the image from registry (same as **image**) or scan the local downloaded image on node (same as **filesystem**). Thanks!
> 
> `rootfs` is working the same as `filesystem`

yes, thanks. We will try with rootfs and see if it works for us.

**Comment by chary1112004** [comment]
> > > > @aarontavio using `trivy.command. = filesystem` is scanning the same image but without downloading it. it use the downloaded image on the node already. but it required higher privileges.
> > > > see here the differences between [fs vs. rootfs support](https://aquasecurity.github.io/trivy/v0.24.4/vulnerability/detection/language/)
> > > 
> > > 
> > > @chary1112004 see this comment
> > 
> > 
> > @chen-keinan I could see the rootfs also scan the same what image scan. What I would like to ask that is **rootfs** download the image from registry (same as **image**) or scan the local downloaded image on node (same as **filesystem**). Thanks!
> 
> `rootfs` is working the same as `filesystem`

We have switched to `rootfs` and it shows error in scan job. As `rootfs` only scan images in nodes so it should no need to pull image.

```
  Warning  Failed             96s (x3 over 2m3s)  kubelet            Error: ErrImageNeverPull
  Warning  ErrImageNeverPull  96s (x3 over 2m3s)  kubelet            Container image "registry.gitlab.com/gitlab-org/ci-cd/gitlab-runner-ubi-images/gitlab-runner-helper-ocp:v16.6.1" is not present with pull policy of Never
  Warning  Failed             96s (x3 over 2m3s)  kubelet            Error: ErrImageNeverPull
  Warning  ErrImageNeverPull  96s (x3 over 2m3s)  kubelet            Container image "registry.gitlab.com/gitlab-org/ci-cd/gitlab-runner-ubi-images/gitlab-runner-helper-ocp:v16.6.1" is not present with pull policy of Never
```

**Comment by chary1112004** [comment]
@chen-keinan I also saw there is option value file **offlineScan: false** https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L373.

Could you please explain how it works for enable **offlineScan:true** for the case **command: image**?

Thanks!

**Comment by chen-keinan** [comment]
> @chen-keinan I also saw there is option value file **offlineScan: false** https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L373.
> 
> Could you please explain how it works for enable **offlineScan:true** for the case **command: image**?
> 
> Thanks!
this flag is if you want to scan `pom.xml` dependencies, you need to specify `--offline-scan` since Trivy tries to issue API requests for scanning Java applications by default.

**Comment by chary1112004** [comment]
@chen-keinan thanks, if we use `rootfs` and `filesystem` then it is mandatory the `new pod` and `scan pod` have to same node? Is it right?
If that is right then trivy operator knows the way to locate `scan pod` to the node of `new pod`? Is it right?

Thanks!

**Comment by chen-keinan** [comment]
> @chen-keinan thanks, if we use `rootfs` or `filesystem` then it is mandatory the `new pod` and `scan pod` have to same node? Is it right? If that is right then trivy operator knows the way to locate `scan pod` to the node of `new pod`? Is it right?
> 
> Thanks!

yes , it schedule it on [specific node](https://github.com/aquasecurity/trivy-operator/blob/main/pkg/plugins/trivy/filesystem.go#L493)

**Comment by chary1112004** [comment]
> > @chen-keinan thanks, if we use `rootfs` or `filesystem` then it is mandatory the `new pod` and `scan pod` have to same node? Is it right? If that is right then trivy operator knows the way to locate `scan pod` to the node of `new pod`? Is it right?
> > Thanks!
> 
> yes , it schedule it on [specific node](https://github.com/aquasecurity/trivy-operator/blob/main/pkg/plugins/trivy/filesystem.go#L493)

Thank you, it could probably that we have gitlab runner pod (`new pod`) belong to specific node group that contains taints then `scan job` could not locate to nodes of this node group then `scan job` and `new pod` are not same node and it is reason ErrImageNeverPull. We will try to add tolerations while use option use `rootfs` or `filesystem`.

Thanks!

**Comment by chen-keinan** [comment]
@chary1112004 do you still want to keep this issue open ?

**Comment by aarontavio** [comment]
Hi @chen-keinan,

thanks for the interest and the effort.

we could verify that after setting `trivy.command` to `filesystem` or `rootfs` the images are not being downloaded anymore. The scanner job gets `imagePullPolicy: Never` which avoids downloading the image for the scanning container and the pod also gets the `nodeName` property set to the same node where the pod to be scanned is running (which should have the image already downloaded). We also could verify in our metrics that data transfer is reduced. By the way the [design_trivy_file_system_scanner.md](https://github.com/aquasecurity/trivy-operator/blob/68dcb68aab9161dae3ec34cb0ced9f1608462f8a/docs/docs/design/design_trivy_file_system_scanner.md) helped me a lot understand how the local scanning works.

So I guess the data transfer problem is solved.

What we have also experienced, a couple of times already, is when the `trivy.command` is set to `filesystem` or `rootfs` some scanning jobs hang with status: ErrImageNeverPull. The scanning pods get blocked and neither kuberntes or the trivy-operator seem to kill the pods, so the whole scanning process is blocked.

See pod events:
```
Events:
   Type     Reason             Age                    From     Message
   ----     ------             ----                   ----     -------
   Normal   Pulled             5m56s                  kubelet  Container image "ghcr.io/aquasecurity/trivy:0.48.2" already present on machine
   Normal   Created            5m56s                  kubelet  Created container 083aae52-c1ad-429f-b9bb-eedd4bd058bc
   Normal   Started            5m56s                  kubelet  Started container 083aae52-c1ad-429f-b9bb-eedd4bd058bc
   Normal   Pulled             5m52s                  kubelet  Container image "ghcr.io/aquasecurity/trivy:0.48.2" already present on machine
   Normal   Created            5m52s                  kubelet  Created container 38d4fc80-6f29-4991-96b3-2e9daa699b40
   Normal   Started            5m52s                  kubelet  Started container 38d4fc80-6f29-4991-96b3-2e9daa699b40
   Warning  Failed             3m52s (x9 over 5m38s)  kubelet  Error: ErrImageNeverPull
   Warning  ErrImageNeverPull  55s (x23 over 5m38s)   kubelet  Container image "confluentinc/cp-kafka:7.5.3" is not present with pull policy of Never
```
This obviously happens whenever the image is not present in the node for some reason. Since the `imagePullPolicy` is set to `Never`. One possible solution could be to create a new user parameter to control the `imagePullPolicy` this way we could set it to `IfNotPresent` and this would guarantee that the scan process never gets blocked. And for us would be ok.

Another option would be that the operator notices this and kills the scanning pod so that the process can continue scanning other images.

Maybe this should be another ticket.

Do you have any idea that can help on this?

**Comment by chen-keinan** [comment]
> 
> Maybe this should be another ticket.
> 
> Do you have any idea that can help on this?

@chary1112004 Please do open another issue (to have the discussion more accurate for community as well) for it and you can close this one if not relevant anymore

**Comment by chary1112004** [comment]
> > Maybe this should be another ticket.
> > Do you have any idea that can help on this?
> 
> @chary1112004 Please do open another issue (to have the discussion more accurate for community as well) for it and you can close this one if not relevant anymore

@chen-keinan yes, I am going to open another ticket for above issue. Thanks for your support!

**Comment by chary1112004** [comment]
> > Maybe this should be another ticket.
> > Do you have any idea that can help on this?
> 
> @chary1112004 Please do open another issue (to have the discussion more accurate for community as well) for it and you can close this one if not relevant anymore

@chen-keinan yes, I am going to open another ticket for above issue. Thanks for your support!
